ambari-server/src/main/java/org/apache/ambari/server/security/encryption/PasswordPropertiesEncryptor.java

```diff
@@ -0,0 +1,150 @@
+    ambariEventPublisher.register(this);
+  }
+
+  /**
+   * It is a must that any caller of this method adds the following additional
+   * information on the following indexes in <code>additionalInfo</code> with the
+   * following types:
+   * <ul>
+   * <li>[0] - <code>org.apache.ambari.server.state.Cluster</code> (the cluster
+   * whose properties should be scanned for PASSWORD properties)</li>
+   * <li>[1] - <code>java.lang.String</code> (to keep properties only that match
+   * this configuration type )</li>
+   * </ul>
+   */
+  @Override
+  public Map<String, String> encryptSensitiveData(Map<String, String> properties, Object... additionaInfo) {
+    try {
+      final Map<String, String> encryptedProperties = new HashMap<>(properties.size());
+      if (properties != null) {
+        final Cluster cluster = (Cluster) additionaInfo[0];
+        final String configType = (String) additionaInfo[1];
+        final Set<String> passwordProperties = getPasswordProperties(cluster, configType);
+        if (CollectionUtils.isNotEmpty(passwordProperties)) {
+          String encryptedValue;
+          for (Map.Entry<String, String> property : properties.entrySet()) {
+            encryptedValue = passwordProperties.contains(property.getKey()) && !isEncryptedPassword(property.getValue())
+                ? encryptAndDecoratePropertyValue(property.getValue())
+                : property.getValue();
+            encryptedProperties.put(property.getKey(), encryptedValue);
+          }
+        }
+      }
+      return encryptedProperties;
+    } catch (Exception e) {
+      throw new AmbariRuntimeException("Error while encrypting sensitive data", e);
+    }
+  }
+
+  private boolean isEncryptedPassword(String password) {
+    return password != null && password.startsWith(ENCRYPTED_PROPERTY_PREFIX); // assuming previous encryption by this class
+  }
+
+  private Set<String> getPasswordProperties(Cluster cluster, String configType) throws AmbariException {
+    final long clusterId = cluster.getClusterId();
+    if (!clusterPasswordProperties.containsKey(clusterId)) {
+      clusterPasswordProperties.put(clusterId, new ConcurrentHashMap<>());
+    }
+    if (!clusterPasswordProperties.get(clusterId).containsKey(configType)) {
+      clusterPasswordProperties.get(clusterId).put(configType, cluster.getConfigPropertiesTypes(configType).get(PropertyType.PASSWORD));
+    }
+
+    return clusterPasswordProperties.get(clusterId).get(configType);
+  }
+
+  private String encryptAndDecoratePropertyValue(String propertyValue) throws Exception {
+    final String encrypted = encryptionService.encrypt(propertyValue, TextEncoding.BIN_HEX);
+    return String.format(ENCRYPTED_PROPERTY_SCHEME, encrypted);
+  }
+
+  @Override
+  public Map<String, String> decryptSensitiveData(Map<String, String> properties, Object... additionalInfo) {
+    final Map<String, String> decryptedProperties = new HashMap<>(properties.size());
```

Comment:
Are not we replacing here an instance of ConcurrentHashMap with an instance of a plain HashMap (that is unsafe for concurrent use)? I think we access and modify cluster properties from different threads

Reply:
Those properties are generated by ConfigFactory after they have been parsed out from the request: https://github.com/apache/ambari/blob/trunk/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ConfigurationResourceProvider.java#L139
As you can see this is a simple HashMap. Which is ok since we guard the actual Config creation with locks and can not be happening that we have multiple properties map for the same version.

Moreove, those properties are being queried in ConfigImpl.getProperties(), which did simply wrapped the internal properties map in a HashMap in previous versions and does it here too.